drivers/azure/azureutil/securityrules.go

```diff
@@ -0,0 +1,310 @@
+	for _, rule := range *rules {
+		if strings.ToLower(fmt.Sprintf("%s", rule.Direction)) != "inbound" {
+			// Ignore outbound rules
+			continue
+		}
+		priority := to.Int32(rule.Priority)
+		if priority == sshPriority {
+			name := to.String(rule.Name)
+			if name == sshPolicyName {
+				// SSH rule already exists
+				return nil
+			}
+			return fmt.Errorf("Cannot set SSH rule as priority %d is already taken by rule %s", sshPriority, name)
+		}
+	}
+	// If not, add the rule
+	*rules = append(*rules, network.SecurityRule{
+		Name: to.StringPtr(sshPolicyName),
+		SecurityRulePropertiesFormat: &network.SecurityRulePropertiesFormat{
+			Description:              to.StringPtr("Allow ssh from public Internet"),
+			SourceAddressPrefix:      to.StringPtr("*"),
+			DestinationAddressPrefix: to.StringPtr("*"),
+			SourcePortRange:          to.StringPtr("*"),
+			DestinationPortRange:     to.StringPtr(fmt.Sprintf("%d", sshPort)),
+			Access:                   network.SecurityRuleAccessAllow,
+			Direction:                network.SecurityRuleDirectionInbound,
+			Protocol:                 network.SecurityRuleProtocolTCP,
+			Priority:                 to.Int32Ptr(sshPriority),
+		},
+	})
+	return nil
+}
+
+func addDockerRule(rules *[]network.SecurityRule, dockerPort int) error {
+	// Check if any existing rule has the same priority as the Docker rule or check any conflicts
+	for _, rule := range *rules {
+		if strings.ToLower(fmt.Sprintf("%s", rule.Direction)) != "inbound" {
+			// Ignore outbound rules
+			continue
+		}
+		priority := to.Int32(rule.Priority)
+		if priority == dockerPriority {
+			name := to.String(rule.Name)
+			if name == dockerPolicyName {
+				// Docker rule already exists
+				return nil
+			}
+			return fmt.Errorf("Cannot set Docker rule as priority %d is already taken by rule %s", dockerPriority, name)
+		}
+		if priority < dockerPriority {
+			// If the priority is less than the docker priority, we should check if there's a conflict
+			err := checkIfDockerConflict(rule, dockerPort)
+			if err != nil {
+				return err
+			}
+		}
+	}
+	// If not, add the rule
+	*rules = append(*rules, network.SecurityRule{
+		Name: to.StringPtr(dockerPolicyName),
+		SecurityRulePropertiesFormat: &network.SecurityRulePropertiesFormat{
+			Description:              to.StringPtr("Allow docker engine access (TLS-protected)"),
+			SourceAddressPrefix:      to.StringPtr("*"),
+			DestinationAddressPrefix: to.StringPtr("*"),
+			SourcePortRange:          to.StringPtr("*"),
+			DestinationPortRange:     to.StringPtr(fmt.Sprintf("%d", dockerPort)),
+			Access:                   network.SecurityRuleAccessAllow,
+			Direction:                network.SecurityRuleDirectionInbound,
+			Protocol:                 network.SecurityRuleProtocolTCP,
+			Priority:                 to.Int32Ptr(dockerPriority),
+		},
+	})
+	return nil
+}
+
+func syncOpenPorts(rules *[]network.SecurityRule, openPorts []string) error {
```

Comment:
I am concerned we are going to modify an existing NSG doing this, is that the intention here?

Reply:
More specifically, we are deleting things from the NSG

Reply:
yes, as detailed in https://github.com/rancher/machine/pull/77#discussion_r420483638 the intention is to modify the existing NSG by adding in SSH, Docker, and Custom Open Ports to it.

We never delete any existing security rules unless they follow the format `Port%s-%sAllowAny` (which is the format used by Rancher for adding in Open Ports). To determine if open port security rules need to be deleted / added, we follow the following procedure:
- Get all rules that follow the `Port%s-%sAllowAny` defined in the security group and mark all of them as to be deleted
- Parse through the requested open ports. If they exist already, remove the mark to delete. Otherwise, mark them as new rules to be added.
- Delete any remaining security rules (as they no longer exist in the openPorts list provided; this indicates that the user changed the openPorts on the node template)
- Add the new rules.

Reply:
As of https://github.com/rancher/machine/pull/77#issuecomment-625564555, we no longer modify an existing NSG that is not managed by Rancher.